pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -3589,6 +3625,79 @@ func pruneRegistrationEntries(tx *gorm.DB, expiresBefore time.Time, logger logru
 	return nil
 }
 
+func createRegistrationEntryEvent(tx *gorm.DB, entryID string) error {
+	if !fflag.IsSet(fflag.FlagEventsBasedCache) {
+		return nil
+	}
+
+	newRegisteredEntryEvent := RegisteredEntryEvent{
+		EntryID: entryID,
+	}
+
+	if err := tx.Create(&newRegisteredEntryEvent).Error; err != nil {
+		return sqlError.Wrap(err)
+	}
+
+	return nil
+}
+
+func listRegistrationEntriesEvents(tx *gorm.DB, req *datastore.ListRegistrationEntriesEventsRequest) (*datastore.ListRegistrationEntriesEventsResponse, error) {
+	if req.Pagination != nil && req.Pagination.PageSize == 0 {
+		return nil, status.Error(codes.InvalidArgument, "cannot paginate with pagesize = 0")
+	}
+
+	p := req.Pagination
+	var err error
+	if p != nil {
+		tx, err = applyPagination(p, tx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var events []RegisteredEntryEvent
+	if err := tx.Find(&events, "id > ?", req.GreaterThanEventID).Error; err != nil {
+		return nil, sqlError.Wrap(err)
+	}
+
+	if p != nil {
+		p.Token = ""
+		// Set token only if page size is the same as events len
+		if len(events) > 0 {
+			lastEntry := events[len(events)-1]
+			p.Token = fmt.Sprint(lastEntry.ID)
+		}
+	}
+
+	resp := &datastore.ListRegistrationEntriesEventsResponse{
+		Pagination: p,
+	}
+	for _, event := range events {
+		resp.EntryIDs = append(resp.EntryIDs, event.EntryID)
+	}
+	if len(events) > 0 {
+		resp.FirstEventID = events[0].ID
+	}
+
+	return resp, nil
+}
+
+func pruneRegistrationEntriesEventsSQLite(tx *gorm.DB, olderThan time.Duration) error {
+	if err := tx.Where(fmt.Sprintf("created_at < datetime('now', 'localtime', '-%d seconds')", int64(olderThan.Seconds()))).Delete(&RegisteredEntryEvent{}).Error; err != nil {
```

Comment:
I couldnt get the `?` syntax working here so thats why its using a `Sprintf`.

Reply:
This doesn't provide an injection attack surface, as `olderThan` is strongly typed to be a `time.Duration` and the `Sprintf` doesn't permit un=sanitized strings to be formatted into the query.

Reply:
This is basically "ok" don't take my comment above as a blocker.  It's a comment that the above solution is fine.

Reply:
? syntax would be more consistent but it doesnt work with the single quotes.